cirq/contrib/qasm_import/qasm_test.py

```diff
@@ -11,24 +11,77 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+from typing import List
+
 import numpy as np
 
 import cirq
 import cirq.testing as ct
 
 
-# TODO as we grow the language, we'll add more complex examples here
-def test_consistency_with_qasm_output():
-    a, b, c, d = [cirq.NamedQubit('q_{}'.format(i)) for i in range(4)]
+def test_consistency_with_qasm_output_and_qiskit():
+    qubits = [cirq.NamedQubit('q_{}'.format(i)) for i in range(4)]
+    a, b, c, d = qubits
     circuit1 = cirq.Circuit.from_ops(
         cirq.Rx(np.pi / 2).on(a),
         cirq.Ry(np.pi / 2).on(b),
         cirq.Rz(np.pi / 2).on(b),
+        cirq.X.on(a),
+        cirq.Y.on(b),
+        cirq.Z.on(c),
+        cirq.H.on(d),
+        cirq.S.on(a),
+        cirq.T.on(b),
+        cirq.S.on(c)**-1,
+        cirq.T.on(d)**-1,
+        cirq.X.on(d)**0.125,
+        cirq.TOFFOLI.on(a, b, c),
+        cirq.CSWAP.on(d, a, b),
+        cirq.SWAP.on(c, d),
+        cirq.CX.on(a, b),
+        cirq.ControlledGate(cirq.Y).on(c, d),
+        cirq.CZ.on(a, b),
+        cirq.ControlledGate(cirq.H).on(b, c),
         cirq.IdentityGate(1).on(c),
         cirq.circuits.qasm_output.QasmUGate(1.0, 2.0, 3.0).on(d),
     )
 
-    qasm1 = cirq.qasm(circuit1)
+    qasm = cirq.qasm(circuit1)
+
+    circuit2 = cirq.contrib.qasm_import.qasm.QasmCircuitParser().parse(qasm)
+
+    cirq_unitary = cirq.unitary(circuit2)
+    ct.assert_allclose_up_to_global_phase(cirq_unitary,
+                                          cirq.unitary(circuit1),
+                                          atol=1e-8)
+
+    # coverage: ignore
+    try:
+        # We don't want to require qiskit as a dependency but
+        # if Qiskit is installed, test QASM output against it.
+        import qiskit  # type: ignore
+    except ImportError:
+        return
+
+    result = qiskit.execute(qiskit.load_qasm_string(qasm),
+                            backend=qiskit.Aer.get_backend('unitary_simulator'))
+    qiskit_unitary = result.result().get_unitary()
+    qiskit_unitary = _reorder_indices_of_matrix(
+        qiskit_unitary, list(reversed(range(len(qubits)))))
+
+    cirq.testing.assert_allclose_up_to_global_phase(cirq_unitary,
+                                                    qiskit_unitary,
+                                                    rtol=1e-8,
+                                                    atol=1e-8)
+
 
-    circuit2 = cirq.contrib.qasm_import.qasm.QasmCircuitParser().parse(qasm1)
-    ct.assert_same_circuits(circuit1, circuit2)
+def _reorder_indices_of_matrix(matrix: np.ndarray, new_order: List[int]):
```

Comment:
this was a shameless copy paste from `cirq.circuits.qasm_output_test` - should I pull it out to a test utility in `cirq.testing`?

Reply:
I was thinking something more along the lines of defining the method `assert_parsed_qasm_circuit_agrees_with_qiskit`, and yes sharing it with the output test. Note that this method should not be in `cirq.testing` because it has a dependency on qiskit.

Reply:
Interesting. I would like to clarify the packages/dependency rules before I jump in: 
 
1. `cirq.testing` has `assert_qasm_is_consistent_with_unitary` in consistent_qasm.py which does refer to qiskit which file also contains a copy of `_reorder_indices_of_matrix` (it's also in cirq.circuits.qasm_output_test which has a `test_output_unitary_same_as_qiskit` test closer to what I wanted to do here)
2. I suspect that the rule of not calling contrib from cirq still stands for tests? In that case no qasm import functionality should be called from qasm output / consistency tests under cirq.testing or cirq.circuit

How about: 

1. `consistent_qasm.py` would be still the home for the qiskit related consistency checks. 
2. I would add a new, simpler assertion method `assert_qiskit_parsed_qasm_consistent_with_unitary(qasm, unitary)` that checks the passed in `unitary` against the qiskit parsed version of a passed in `qasm` string. This can then be used in `contrib.qasm_import.qasm_test` as well as in `qasm_output_test`

Reply:
I added a commit implementing this idea:  https://github.com/quantumlib/Cirq/pull/1830/commits/462a069d3782796de41bb8624f48eed12dad374d Let me know what you think!